contracts/modules/v2/Assessment.sol

```diff
@@ -35,238 +27,286 @@ contract Assessment is IAssessment, MasterAwareV2 {
+  returns (uint total, uint withdrawable, uint withdrawableUntilIndex) {
+    Stake memory stake = stakeOf[user];
+    Vote memory vote;
+    Assessment memory assessment;
+    for (uint i = stake.rewardsWithdrawnUntilIndex; i < votesOf[user].length; i++) {
+      vote = votesOf[user][i];
+      assessment = assessments[vote.assessmentId];
+      if (
+        withdrawableUntilIndex == 0 &&
+        assessment.poll.end + config.payoutCooldownDays * 1 days >= block.timestamp
+      ) {
+        withdrawableUntilIndex = i;
+        withdrawable = total;
+      }
+
+      total += assessment.totalReward * vote.tokenWeight /
+        (assessment.poll.accepted + assessment.poll.denied);
+    }
   }
 
-  function withdrawStake(uint96 amount) external override onlyMember {
-    AssessmentVoteLib.withdrawStake(config, nxm, stakeOf, votesOf, amount);
+  /// Withdraws a staker's accumulated rewards
+  /// @dev
+  ///
+  /// @param user        The address of the staker for which the rewards are withdrawn
+  /// @param untilIndex  The index until which the rewards should be withdrawn. Used if a large
+  ///                    number of assessments accumulates and the function doesn't fir in one
+  ///                    block, thus requiring multiple batched transactions.
+  function withdrawReward(address user, uint104 untilIndex) external override
+  returns (uint withdrawn, uint withdrawUntilIndex) {
+    Stake memory stake = stakeOf[user];
+    {
+      uint voteCount = votesOf[user].length;
+      withdrawUntilIndex = untilIndex > 0 ? untilIndex : voteCount;
+      require(
+        untilIndex <= voteCount,
+        "Vote count is smaller that the provided untilIndex"
+      );
+      require(stake.rewardsWithdrawnUntilIndex < voteCount, "No withdrawable rewards");
+    }
+
+    Vote memory vote;
+    Assessment memory assessment;
+    for (uint i = stake.rewardsWithdrawnUntilIndex; i < withdrawUntilIndex; i++) {
+      vote = votesOf[user][i];
+      assessment = assessments[vote.assessmentId];
+      if (assessment.poll.end + config.payoutCooldownDays * 1 days >= block.timestamp) {
+        // Poll is not final
+        withdrawUntilIndex = i;
+        break;
+      }
+
+      withdrawn += assessment.totalReward * vote.tokenWeight /
+        (assessment.poll.accepted + assessment.poll.denied);
+    }
+
+    // This is the index where the next withdrawReward call will start iterating from
+    stakeOf[user].rewardsWithdrawnUntilIndex = uint104(withdrawUntilIndex);
+    ITokenController(getInternalContractAddress(ID.TC)).mint(user, withdrawn);
   }
 
-  function redeemClaimPayout(uint104 id) external override {
-    AssessmentClaimsLib.redeemClaimPayout(
-      config,
-      internalContracts,
-      claims,
-      claimants,
-      id
-    );
+  function withdrawStake(uint96 amount) external override onlyMember {
+    Stake storage stake = stakeOf[msg.sender];
+    require(stake.amount != 0, "No tokens staked");
```

Comment:
check that `amount <= stake.amount` instead
also - does it make sense to check if amount > 0?

Reply:
Unsigned integers implicitly revert in this solidity version when underflowing. At the same time uints can never be <0 so there's no need to check the amount.